### ISSUE TITLE
JavaScript: Objects 'Reflect' and 'Proxy' are defined but empty, which breaks Babel

### ISSUE THREAD

**Issue by rdeforest** [nesbox/TIC-80#2097]
I'm using CoffeeScript to generate JS, Babel to turn it into ES5, and rollup to make `include "libgame"` work for me. Crazy, I know! Anyway...

Babel uses the existence of an object called `Reflect` to decide how to implement the `super()` call, but in the version of duktape TIC-80 uses, this object exists but has no `construct` method. If Reflect didn't exist at all my setup would work fine. In fact, it works fine as long as I don't use `super()`, which will be my workaround. (Multiple inheritance? Ha! You don't even get single inheritance! Deal with it!).

In the interest of providing a good bug report, here's some CoffeeScript to reproduce the issue:

```coffee
log = (args...) ->
  for arg in args
    trace JSON.stringify arg, null, 2

class Parent
  constructor: ->
    log "parent constructor"

class Child extends Parent
  constructor: ->
    super()
    log "child constructor"

kid = new Child

log 'undefined' isnt typeof kid
```

The ES5 that turns into is 149 lines long. I'll provide the whole thing on request, but here are the relevant portions:

```js
 function _isNativeReflectConstruct() {
    if (typeof Reflect === "undefined" || !Reflect.construct) return false;
    if (Reflect.construct.sham) return false;
    if (typeof Proxy === "function") return true;
    try {
      Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {}));
      return true;
    } catch (e) {
      return false;
    }
  }

  function _createSuper(Derived) {
    var hasNativeReflectConstruct = _isNativeReflectConstruct();
    return function _createSuperInternal() {
      var Super = _getPrototypeOf(Derived),
        result;
      if (hasNativeReflectConstruct) {
        var NewTarget = _getPrototypeOf(this).constructor;
        result = Reflect.construct(Super, arguments, NewTarget);
      } else {
        result = Super.apply(this, arguments);
      }
      return _possibleConstructorReturn(this, result);
    };
  }
```

I suppose a case could be made that Babel isn't behaving conservatively enough, and in the strictest sense that's true. My reason for calling this a bug is that there's no reason for these objects to both exists and not provide the functionality their names are associated with. It violates the principle of least surprise.

**Comment by joshgoebel** [comment]
I don't know why they exist, but can you not just undefine them?

**Comment by rdeforest** [comment]
I'll give that a try as soon as I get my janky system working again. :)

**Comment by rdeforest** [comment]
It worked! I don't know why that work-around didn't occur to me. Thank you for the quick response!

**Comment by rdeforest** [comment]
Well, it did work until I uncommented my more complex code. Now I'm getting `identifier 'global' undefined`. Maybe I'll put something in the wiki if I ever get this all working to my expectations.

**Comment by joshgoebel** [comment]
> Well, it did work ... Now I'm getting identifier 'global' undefined.

So sounds like it's STILL working... that error is a *different* error entirely... one thing at a time. :-).

**Comment by rdeforest** [comment]
Yes, that's what I meant. :)

As promised, here's all the magic:

### babel.config.json

```json
{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {},
        "useBuiltIns": "usage",
        "corejs": "3.6.5",
        "forceAllTransforms": true
      }
    ]
  ]
}
```

### rollup.config.mjs
```js
import glob from 'glob';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { babel } from '@rollup/plugin-babel';
import resolve from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';

const config = {
  input: Object.fromEntries(
		glob.sync('src/js/*.js').map(file => [
			// This remove `src/` as well as the file extension from each
			// file, so e.g. src/nested/foo.js becomes nested/foo
			path.relative(
				'src/js',
				file.slice(0, file.length - path.extname(file).length)
			),
			// This expands the relative paths to absolute paths, so e.g.
			// src/nested/foo becomes /project/src/nested/foo.js
			fileURLToPath(new URL(file, import.meta.url))
		])
	),
  output: {
    dir: 'src/es5',
    format: 'iife',
    name: 'TIC',
    strict: false
  },
  plugins: [
    resolve(),
    commonjs({ strictRequires: true }),
    babel({ babelHelpers: 'bundled' })
  ]
};

export default config;
```

### Magical shell incantations

```sh
coffee -b -w -o $yourModernJSDirectory $yourCoffeeScriptDirectory
rollup -c -f iife -i $yourModernJSDirecotry/$yourFile
```

### Code I had to put in my scripts to work around bugs

```coffee
delete globalThis.Proxy
delete globalThis.Reflect
globalThis.TIC = ->
  # whatever you do on ticks
  trace "world"
  exit()
globalThis.BOOT = ->
  # whatever you do to initialize
  trace "hello"
```

### Remaining commentary

I know some of this is redundant, and some of it is probably unnecessary, I just wanted to share what I have while it works. I didn't actually test the CoffeeScript snippet above in the form it's in there. What I tested is a bit more complex.

What you get out of all of this is the ability to split your CoffeeScript source into multiple files, and all the features of modern CoffeeScript 2.7.0 and ECMAScript. I have a 'libgame.coffee' which exports a Game class which sets up some sane defaults like binding 'space' to pause/resume, 'q' to quit and 'd' to dump. It also handles setting up the palette.

OH YEAH the palette. Comments in the original script to get converted into JS comments in the final result, but they're indented and not at the very end, so TIC-80 doesn't respect them, which is understandable. A person could work around that in various ways, I chose to have my Game class have a palette property which is a string of hex that gets poked into 0x3FC0. Obviously the same will go for the rest of the sections like TILES and WAVES, which my library doesn't handle yet.

Another benefit of this setup is that one can turn on minification of the JavaScript to shrink the code down to the bare minimum. I haven't tried this yet. It is also possible to include libraries from npm.com in your little TIC-80 project as long as they don't depend on features of NodeJS or other interpreters. Math libraries and lodash come to mind as likely candidates. Also, when rollup includes those libraries into your project, it will strip out as much of those libraries as it safely can, based on what functions and objects you use. When I first started this experiment, my TIC function kept being left out of the end product because rollup couldn't find it being used anywhere. I had to do a `trace TIC && true` to convince rollup not to excise it.

# In conclusion...

I'm having a blast, thank you for developing TIC-80! I'm satisfied with the results so feel free to close this issue if you want.
